include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,347 @@
+            .bitmap = {},
+            .buffers = {{}, array_data::buffer_type(sizeof(std::int64_t), 0)},
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    template <constant_range_for_array_data ValueRange>
+    array_data make_array_data_for_variable_size_binary_layout(
+        ValueRange&& values,
+        const array_data::bitmap_type& bitmap,
+        std::int64_t offset
+    )
+    {
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const auto create_buffers = [&values]()
+        {
+            using T = std::ranges::range_value_t<ValueRange>;
+            const auto& unwrap_value = [](const T& value)
+            {
+                if constexpr (mpl::is_reference_wrapper_v<T>)
+                {
+                    return value.get();
+                }
+                else
+                {
+                    return value;
+                }
+            };
+
+            std::vector<array_data::buffer_type> buffers(2);
+            buffers[0].resize(sizeof(std::int64_t) * (values.size() + 1), 0);
+            buffers[1].resize(std::accumulate(
+                values.begin(),
+                values.end(),
+                size_t(0),
+                [&unwrap_value](std::size_t res, const auto& s)
+                {
+                    return res + unwrap_value(s).size();
+                }
+            ));
+
+            auto iter = buffers[1].begin();
+            const auto offsets = buffers[0].data<std::int64_t>();
+            size_t i = 0;
+            for (const auto& value : values)
+            {
+                const auto& unwraped_value = unwrap_value(value);
+                SPARROW_ASSERT_TRUE(
+                    std::cmp_less(unwraped_value.size(), std::numeric_limits<std::int64_t>::max())
+                );
+                offsets[i + 1] = offsets[i] + static_cast<std::int64_t>(unwraped_value.size());
+                std::ranges::copy(unwraped_value, iter);
+                std::advance(iter, unwraped_value.size());
+                ++i;
+            }
+            return buffers;
+        };
+
+        using T = std::unwrap_ref_decay_t<std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>>;
+
+        return {
+            .type = data_descriptor(arrow_type_id<T>()),
+            .length = static_cast<array_data::length_type>(values.size()),
+            .offset = offset,
+            .bitmap = bitmap,
+            .buffers = create_buffers(),
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    template <typename V>
+    struct values_and_indexes
+    {
+        template <mpl::constant_range R>
+            requires std::same_as<std::ranges::range_value_t<R>, std::remove_const_t<V>>
+        explicit values_and_indexes(R&& range)
+        {
+            ranges_to_vec_and_indexes<R>(std::forward<R>(range), *this);
```

Comment:
Am I missing something or this function is not declared yet at this point? Shouldnt this fail to compile? I'm not sure if that dpeends on the instantiation point or not XD

Reply:
You right and there is no issue at compilation time, so I didn't separated declaration and implementation

Reply:
I'll assume it's resolved at instantiation time I guess.